packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -218,10 +241,220 @@ export class ProtoArray {
+          latestValidHashIndex = nodeIndex;
+          // We found the block corresponding to latestValidHashIndex, exit the loop
+          break;
+        }
+      }
+
+      // We are trying to be as forgiving as possible here because ideally latestValidHashIndex
+      // should be found in the forkchoice
+      if (latestValidHashIndex >= 0) {
+        this.propagateValidExecutionStatusByIndex(latestValidHashIndex);
+      }
+    } else {
+      // In case of invalidation, ideally:
+      //  i) Find the invalid payload
+      //  ii) Obtain a chain [LVH.child, LVH.child.child, ....., invalid_payload]
+      //  iii) Obtain a chain [Last_known_valid_node,  ...., LVH]
+      //
+      // Mark chain iii) as Valid if LVH is non null but right now LVH can be non null without
+      //  gurranteing chain iii) to be valid: for e.g. in following scenario LVH can be returned
+      //  as any of SYNCING: SYNCING, SYNCING, SYNCING, INVALID (due to simple check)/
+      //  So we currently ignore this chain and hope eventually it gets resolved
+      //
+      // Mark chain ii) as Invalid if LVH is found and non null, else only invalidate invalid_payload
+      // if its in fcU.
+      //
+      const {invalidateFromBlockHash, latestValidExecHash} = execResponse;
+      const invalidateFromIndex = this.indices.get(invalidateFromBlockHash);
+      if (invalidateFromIndex === undefined) {
+        throw Error(`Unable to find invalidateFromBlockHash=${invalidateFromBlockHash} in forkChoice`);
+      }
+      const latestValidHashIndex =
+        latestValidExecHash !== null ? this.getNodeIndexFromLVH(latestValidExecHash, invalidateFromIndex) : null;
+      if (latestValidHashIndex === null) {
+        /**
+         *  If the LVH is null or not found, represented with latestValidHashIndex=undefined,
+         *   then just invalidate the invalid_payload and bug out.
+         *
+         *   Ideally in not found scenario we should invalidate the entire chain upwards, but
+         *   it is possible (and observed in the testnets) that the EL was
+         *
+         *     i) buggy: that the LVH was not really the parent of the invalid block, but on
+         *        some side chain
+         *     ii) lazy: that invalidation was result of simple check and the EL just
+         *         responded with a bogus LVH
+         *
+         *   So we will just invalidate the current payload and let future responses take care
+         *   to be as robust as possible.
+         */
+        this.invalidateNodeByIndex(invalidateFromIndex);
+      } else {
+        this.propagateInValidExecutionStatusByIndex(invalidateFromIndex, latestValidHashIndex, currentSlot);
+      }
+    }
+  }
+
+  private propagateValidExecutionStatusByIndex(validNodeIndex: number): void {
+    let nodeIndex: number | undefined = validNodeIndex;
+    // propagate till we keep encountering syncing status
+    while (nodeIndex !== undefined) {
```

Comment:
This loop should also stop when the parent is already valid as a cases in addition to going until invalid, premerge, or thru all ancestors.